### ISSUE TITLE
TitanBlueprintsGraph not releasing transactions?

### ISSUE THREAD

**Issue by smalyshev** [thinkaurelius/titan#914]
It looks like TitanBlueprintsGraph class does not release transactions it puts into openTx map.

I was porting data import code from Titan 0.5.2 to 0.9 and I've encountered OutOfMemory errors. Looking into the memory dump, I've discovered 16 StandardTitanTx objects held by ConcurrentHashMap in StandardTitanGraph object. I suspect those are transactions in openTx map in TitanBlueprintsGraph:

https://github.com/thinkaurelius/titan/blob/0.9.0-M1/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/tinkerpop/TitanBlueprintsGraph.java#L64

Looking at the code, openTx is declared private and only two references to it are:

https://github.com/thinkaurelius/titan/blob/0.9.0-M1/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/tinkerpop/TitanBlueprintsGraph.java#L85

where new transaction gets put into the map, and:
https://github.com/thinkaurelius/titan/blob/0.9.0-M1/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/tinkerpop/TitanBlueprintsGraph.java#L98

where the graph is closed and the map is cleared. But when transactions are committed or rolled back I see no change to openTx and no reference to openTx in other code at all. Am I missing something or indeed openTx keeps them forever until the graph is closed? If so, it seems to be what is causing the memory leak - the ConcurrentHashMap seems to consume 2/3 of the heap and is never cleaned until the process goes OOM.

**Comment by dalaro** [comment]
I agree with your analysis on 0.9.0-M1.  This looks like an unintended reference retention bug in 0.9.0-M1.

I checked titan05 for the same issue.  I think that titan05 is unaffected.  TitanBlueprintsGraph's 0.5.3-SNAPSHOT counterpart removes a thread-bound transaction from the openTx map when the transaction either
- commits: https://github.com/thinkaurelius/titan/blob/titan05/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/blueprints/TitanBlueprintsGraph.java#L63
- rolls back: https://github.com/thinkaurelius/titan/blob/titan05/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/blueprints/TitanBlueprintsGraph.java#L77

I suspect these removals got lost in the TP2-to-TP3 migration that took place on titan09.

**Comment by dalaro** [comment]
I removed the `openTx` map in f25283628168a4888ce46f38e6cf7cecb2794c05.  In that commit, I also made `StandardTitanGraph.commit` throw an exception when `MANUAL`-mode transactions are left hanging.  However, I threw the wrong exception type.  TP3 specifies IllegalArgumentException where I threw TitanException, and there's even a TP3 test to check for this.  Fixed type in 24c1ef84913e4fd79631f2c26931187d22c60d7b.
